imbi/automations.py

```diff
@@ -173,3 +188,99 @@ async def _get_gitlab_parent(self, project: Project) -> dict:
                             project.namespace.gitlab_group_name,
                             project.project_type.gitlab_project_prefix)
         return gitlab_parent
+
+
+class GitLabInitialCommitAutomation(Automation):
+    def __init__(self,
+                 automation_settings: dict,
+                 project_id: int,
+                 cookie_cutter: str,
+                 user: 'imbi.user.User',
+                 db: sprockets_postgres.PostgresConnector):
+        super().__init__(user, db)
+        self._cookie_cutter_name = cookie_cutter
+        self._imbi_project_id = project_id
+
+        self._cookie_cutter: typing.Union[CookieCutter, None] = None
+        self._gitlab_project_info: typing.Union[dict, None] = None
+        self._project: typing.Union[Project, None] = None
+        self._token: typing.Union[imbi.integrations.IntegrationToken,
+                                  None] = None
+
+    async def prepare(self) -> typing.List[str]:
+        self._project = await self._get_project(self._imbi_project_id)
+        self._token = await self._get_gitlab_token()
+        self._cookie_cutter = await self._get_cookie_cutter(
+            self._cookie_cutter_name)
+
+        if not self._has_error() and (self._project.project_type.id !=
+                                      self._cookie_cutter.project_type_id):
+            self._add_error(
+                'Cookie Cutter {} is not available for Project Type {}',
+                self._cookie_cutter.name, self._project.project_type.slug)
+        elif self._project.gitlab_project_id is None:
+            self._add_error('GitLab project does not exist for {}',
+                            self._project.slug)
+        else:
+            self._gitlab = gitlab.GitLabClient(self._token)
+            self._gitlab_project_info = await self._gitlab.fetch_project(
+                self._project.gitlab_project_id)
+            if not self._gitlab_project_info:
+                self._add_error('GitLab project id {} does not exist',
```

Comment:
I want to make sure I understand something. My initial thought when I read this error message was "What if fetch_project gets a 500 response? That doesn't necessarily mean that the project does not exist". I went spelunking through the code and saw that fetch_project will raise an InternalServerError which looks like it bubbles up via ProblemDetails to the Tornado HTTPError which will return an error response to the user. So for that reason, we know if the code has made it to this point then it was a 404 error?